File: src/backend/llm/keycheck.py
import os
import requests
from src.backend.utils.logger import logger
from apscheduler.schedulers.background import BackgroundScheduler

class OpenRouterCreditsChecker:
    """Manages OpenRouter API credits checking"""
    
    def __init__(self):
        self.credits_available = True
        self.scheduler = None
        
    def check_credits(self):
        """Check if we have credits available on OpenRouter"""
        api_key = os.getenv("OPENROUTER_API_KEY")
        base_url = os.getenv("LLM_BASE_URL", "https://openrouter.ai/api/v1")
        
        if not api_key:
            logger.error("OPENROUTER_API_KEY not found")
            self.credits_available = False
            return False
            
        try:
            response = requests.get(
                f"{base_url}/auth/key",
                headers={"Authorization": f"Bearer {api_key}"},
                timeout=5.0
            )
            
            # If we get a 401 or any error response, we're out of credits
            if response.status_code != 200:
                logger.warning(f"OpenRouter credits check failed: {response.status_code}")
                self.credits_available = False
                return False
                
            self.credits_available = True
            return True
            
        except Exception as e:
            logger.error(f"Error checking OpenRouter credits: {str(e)}")
            self.credits_available = False
            return False
    
    def start(self):
        """Start the credits checker"""
        logger.info("Running initial OpenRouter credits check")
        self.check_credits()
        
        # Setup scheduler for periodic checks
        self.scheduler = BackgroundScheduler()
        self.scheduler.add_job(self.check_credits, 'interval', minutes=5, id='credits_check')
        self.scheduler.start()
        
        logger.info("OpenRouter credits check scheduler started")
    
    def stop(self):
        """Stop the credits checker"""
        if self.scheduler:
            self.scheduler.shutdown()
            logger.info("OpenRouter credits check scheduler stopped")
    
    @property
    def has_credits(self):
        """Check if credits are available"""
        return self.credits_available

# Create a global instance
credits_checker = OpenRouterCreditsChecker() 

File: src/backend/llm/keycheck.ts
import axios from 'axios';
import { config } from '../config';
import { logger } from '../utils/logger';

class OpenRouterCreditsChecker {
  private creditsAvailable: boolean = true;
  private checkInterval: NodeJS.Timer | null = null;
  private readonly CHECK_INTERVAL = 5 * 60 * 1000; // 5 minutes in milliseconds

  async checkCredits(): Promise<boolean> {
    if (!config.llm.primary.apiKey) {
      logger.error('OPENROUTER_API_KEY not found');
      this.creditsAvailable = false;
      return false;
    }

    try {
      const response = await axios.get(
        `${config.llm.primary.baseUrl}/auth/key`,
        {
          headers: {
            'Authorization': `Bearer ${config.llm.primary.apiKey}`
          },
          timeout: 5000 // 5 seconds
        }
      );

      // If we get any error response, we're out of credits
      if (response.status !== 200) {
        logger.warn(`OpenRouter credits check failed: ${response.status}`);
        this.creditsAvailable = false;
        return false;
      }

      this.creditsAvailable = true;
      return true;

    } catch (error) {
      logger.error('Error checking OpenRouter credits:', error);
      this.creditsAvailable = false;
      return false;
    }
  }

  start(): void {
    logger.info('Running initial OpenRouter credits check');
    this.checkCredits();

    // Setup interval for periodic checks
    this.checkInterval = setInterval(
      () => this.checkCredits(),
      this.CHECK_INTERVAL
    );

    logger.info('OpenRouter credits check scheduler started');
  }

  stop(): void {
    if (this.checkInterval) {
      clearInterval(this.checkInterval);
      this.checkInterval = null;
      logger.info('OpenRouter credits check scheduler stopped');
    }
  }

  get hasCredits(): boolean {
    return this.creditsAvailable;
  }
}

// Create a global instance
export const creditsChecker = new OpenRouterCreditsChecker(); 